roachtest: sqlsmith/setup=empty/setting=no-mutations failed

[(roachtest).sqlsmith/setup=empty/setting=no-mutations failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=2907612&tab=buildLog) on [release-21.1@f3b14a95b781f57ac53a472b9979270b0ecbedfb](https://github.com/cockroachdb/cockroach/commits/f3b14a95b781f57ac53a472b9979270b0ecbedfb):

```
The test failed on branch=release-21.1, cloud=gce:
test artifacts and logs in: /home/agent/work/.go/src/github.com/cockroachdb/cockroach/artifacts/sqlsmith/setup=empty/setting=no-mutations/run_1
	sqlsmith.go:206,sqlsmith.go:237,test_runner.go:767: error: pq: internal error: unsupported comparison: date to timestamp
		stmt:
		SELECT
			tab_3542.col_6057 AS col_6061, tab_3543.col_6058 AS col_6062
		FROM
			(VALUES (NULL)) AS tab_3542 (col_6057)
			INNER JOIN (VALUES (ARRAY[]:::TIMESTAMP[])) AS tab_3543 (col_6058) ON
					(((SELECT '-infinity':::DATE AS col_6060 FROM (VALUES ('{"V/.?;8^,jY": [[[true]], 0.7446110091390192], "foobar": {"foo": [], "hRL": {"9o+": "4Vv)", "b": [{}, {}, {}], "rrR8(DQ": null}}}':::JSONB)) AS tab_3544 (col_6059) LIMIT 1:::INT8)::DATE - '49 years 5 mons 927 days 18:38:51.109843':::INTERVAL::INTERVAL)::TIMESTAMP < '1971-01-17 22:36:35.000855':::TIMESTAMP)
		ORDER BY
			tab_3542.col_6057 DESC, tab_3543.col_6058
		LIMIT
			67:::INT8;

```

<details><summary>More</summary><p>


Artifacts: [/sqlsmith/setup=empty/setting=no-mutations](https://teamcity.cockroachdb.com/viewLog.html?buildId=2907612&tab=artifacts#/sqlsmith/setup=empty/setting=no-mutations)

[See this test on roachdash](https://roachdash.crdb.dev/?filter=status%3Aopen+t%3A.%2Asqlsmith%2Fsetup%3Dempty%2Fsetting%3Dno-mutations.%2A&sort=title&restgroup=false&display=lastcommented+project)
<sub>powered by [pkg/cmd/internal/issues](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)</sub></p></details>

--- a/pkg/sql/logictest/testdata/logic_test/datetime
+++ b/pkg/sql/logictest/testdata/logic_test/datetime
@@ -1756,3 +1756,10 @@ NULL                             NULL
 
 statement ok
 RESET TIME ZONE
+
+# Regression test for panicking when comparing an infinite date coming from a
+# subquery to a timestamp (#64015).
+query B
+SELECT ((SELECT '-infinity'::DATE)) < '2021-04-21':::TIMESTAMP
+----
+true

--- a/pkg/sql/sem/tree/datum.go
+++ b/pkg/sql/sem/tree/datum.go
@@ -1866,12 +1866,12 @@ func (d *DDate) Next(_ *EvalContext) (Datum, bool) {
 
 // IsMax implements the Datum interface.
 func (d *DDate) IsMax(_ *EvalContext) bool {
-	return d.Date == pgdate.PosInfDate
+	return d.PGEpochDays() == pgdate.PosInfDate.PGEpochDays()
 }
 
 // IsMin implements the Datum interface.
 func (d *DDate) IsMin(_ *EvalContext) bool {
-	return d.Date == pgdate.NegInfDate
+	return d.PGEpochDays() == pgdate.NegInfDate.PGEpochDays()
 }
 
 // Max implements the Datum interface.
@@ -2264,7 +2264,49 @@ func timeFromDatumForComparison(ctx *EvalContext, d Datum) (time.Time, bool) {
 	}
 }
 
+type infiniteDateComparison int
+
+const (
+	// Note: the order of the constants here is important.
+	negativeInfinity infiniteDateComparison = iota
+	finite
+	positiveInfinity
+)
+
+func checkInfiniteDate(ctx *EvalContext, d Datum) infiniteDateComparison {
+	if _, isDate := d.(*DDate); isDate {
+		if d.IsMax(ctx) {
+			return positiveInfinity
+		}
+		if d.IsMin(ctx) {
+			return negativeInfinity
+		}
+	}
+	return finite
+}
+
+// compareTimestamps takes in two time-related datums and compares them as
+// timestamps while paying attention to time zones if needed. It returns -1, 0,
+// or +1 for "less", "equal", and "greater", respectively.
+//
+// Datums are allowed to be one of DDate, DTimestamp, DTimestampTZ, DTime,
+// DTimeTZ. For all other datum types it will panic; also, comparing two DDates
+// is not supported.
 func compareTimestamps(ctx *EvalContext, l Datum, r Datum) int {
+	leftInf := checkInfiniteDate(ctx, l)
+	rightInf := checkInfiniteDate(ctx, r)
+	if leftInf != finite || rightInf != finite {
+		// At least one of the datums is an infinite date.
+		if leftInf != finite && rightInf != finite {
+			// Both datums cannot be infinite dates at the same time because we
+			// wouldn't use this method.
+			panic(errors.AssertionFailedf("unexpectedly two infinite dates in compareTimestamps"))
+		}
+		// Exactly one of the datums is an infinite date and another is a finite
+		// datums (not necessarily a date). We can just subtract the returned
+		// values to get the desired result for comparison.
+		return int(leftInf - rightInf)
+	}
 	lTime, lOk := timeFromDatumForComparison(ctx, l)
 	rTime, rOk := timeFromDatumForComparison(ctx, r)
 	if !lOk || !rOk {

--- a/pkg/sql/sem/tree/datum_test.go
+++ b/pkg/sql/sem/tree/datum_test.go
@@ -121,6 +121,18 @@ func TestCompareTimestamps(t *testing.T) {
 			right:    NewDTimeTZFromOffset(timeofday.New(5, 0, 0, 0), pacificTimeZone),
 			expected: 1,
 		},
+		{
+			desc:     "positive infinite date",
+			left:     dMaxDate,
+			right:    MakeDTime(timeofday.New(12, 0, 0, 0)),
+			expected: 1,
+		},
+		{
+			desc:     "negative infinite date",
+			left:     dMinDate,
+			right:    MakeDTime(timeofday.New(12, 0, 0, 0)),
+			expected: -1,
+		},
 	}
 
 	for _, tc := range testCases {
@@ -137,6 +149,8 @@ func TestCompareTimestamps(t *testing.T) {
 			},
 		)
 	}
+
+	assert.Panics(t, func() { compareTimestamps(nil /* ctx */, dMaxDate, dMinDate) })
 }
 
 func TestCastStringToRegClassTableName(t *testing.T) {

